File: src/Kotlin/Killer.kt
import java.io.File
import java.sql.Timestamp
import java.util.*
import java.util.stream.IntStream
import kotlin.system.exitProcess

object Killer {

    private var params = HashMap<Int, List<Int>>()
    private var combos = Array(81) {Array(81) {false} }

    @JvmStatic
    fun main(args: Array<String>) {
        println(Timestamp(System.currentTimeMillis()))
        val intVals: List<List<Int>> = File("/home/sandy/puzzle.csv").readLines().map { it -> it.split(",").map { it.toInt() } }
        intVals.forEach { line -> params[line[line.size - 1]] = line }
        println("params: $params")
        for(i:Int in 0 until 81) {
            for (j: Int in 0 until i) {
                combos[i][j] = (sameRow(i, j) || sameCol(i, j) || sameBlock(i, j))
            }
        }
        val newBoard = IntArray(81)
        recursiveCheck(newBoard)
    }

    private fun getSelectedParam(position: Int): List<Int>? {
        return if (params[position] != null)
            params[position]
        else
        //return arrayListOf<Int>()
            emptyList()
    }

    private fun sameRow(i: Int, j: Int): Boolean {
        //return i / 9 shr 0 == j / 9 shr 0
        return (i/9) == (j/9)
    }

    private fun sameCol(k: Int, l: Int): Boolean {
        return (k - l) % 9 == 0
    }

    private fun sameBlock(m: Int, n: Int): Boolean {
        return m / 27 == n / 27 && m % 9 / 3 == n % 9 / 3
    }

    private fun paramTotal(param: List<Int>?, aRow: IntArray, value: Int): Int? {
        val subParam = param?.subList(1, param.size - 1)
        return subParam?.stream()?.reduce(value) { subtotal: Int, element: Int? ->
            subtotal + aRow[element!!]
        }
    }

    private fun checkRow(aRow: IntArray, position: Int, value: Int): Boolean {
        val selectedParam = getSelectedParam(position)
        return if (selectedParam.isNullOrEmpty()) {
            true
        } else {
            val total = paramTotal(selectedParam, aRow, value)
            total == selectedParam[0]
        }
    }

    private fun recursiveCheck(inBoard: IntArray): Boolean {
        val pos = IntStream.range(0, inBoard.size).filter { i: Int -> inBoard[i] == 0 }.findFirst().orElse(-1)
        if (pos == -1) {
            println("results:" + inBoard.contentToString())
            println(Timestamp(System.currentTimeMillis()))
            // return true;
            exitProcess(1)
        }
        val excludedNumbers: MutableSet<Int> = HashSet()
        for (x in 0 until pos) {
            if (combos[pos][x]) {
                excludedNumbers.add(inBoard[x])
            }
        }
        for (y in 1..9) {
            if (!excludedNumbers.contains(y)) {
                if (checkRow(inBoard, pos, y)) {
                    val newBoard = inBoard.clone()
                    newBoard[pos] = y
                    recursiveCheck(newBoard)
                }
            }
        }
        return true
    }
}

File: src/Dart/killer.dart
import 'dart:collection';
import 'dart:io';

var board = List.filled(81, 0);
var related = List.generate(81, (i) => List.filled(81, ''));
var parms = HashMap<int, List<int>>();

void getParms() {
  File('/home/simon/Killer/C/killer/puzzle.csv')
      .readAsLinesSync()
      .forEach((el) {
    var a = el.split(',').map(int.parse).toList();
    parms[a[a.length - 1]] = a;
  });
}

bool checkRules(List board, int i, int m) {
  var p = parms[i];
  if (p == null) {
    return true;
  }

  var sum = m;

  for (var x = 1; x < p.length; x++) {
    if (p[x] != i) {
      if (board[p[x]] == 0) {
        return true;
      }
      sum += board[p[x]];
    }
  }

  if (sum == p[0]) {
    return true;
  } else {
    return false;
  }
}

bool checkBoardValid(List board, int i, int m) {
  bool sameRow(int i, int j) {
    return i ~/ 9 == j ~/ 9;
  }

  bool sameCol(int i, int j) {
    return (i - j) % 9 == 0;
  }

  bool sameBlock(int i, int j) {
    return ((i ~/ 27 == j ~/ 27) && (i % 9 ~/ 3 == j % 9 ~/ 3));
  }

  for (var j = 0; j < i; j++) {
    if (related[i][j] == '') {
      if (sameRow(i, j) || sameCol(i, j) || sameBlock(i, j)) {
        related[i][j] = 'Y';
        related[j][i] = 'Y';
      } else {
        related[i][j] = 'N';
        related[j][i] = 'N';
      }
    }
    if (related[i][j] == 'Y') if (board[j] == m) return false;
  }

  return true;
}

void recursiveCheck(List board) {
  var i = board.indexOf(0);
  if (i == -1) {
    print(board.toString());
    exit(0);
  }

  for (var m = 1; m < 10; m++) {
    if (checkBoardValid(board, i, m)) {
      if (checkRules(board, i, m)) {
        var newBoard = List.of(board);
        newBoard[i] = m;
        recursiveCheck(newBoard);
      }
    }
  }
}

void main(List<String> arguments) {
  print('Started...');
  getParms();
  recursiveCheck(board);
}
